data/tutorials/ds_05_seq.md

```diff
@@ -0,0 +1,317 @@
+```ocaml
+# Seq.ints 0 |> Seq.take 43 |> List.of_seq;;
+- : int list =
+[0; 1; 2; 3; 4; 5; 6; 7; 8; 9; 10; 11; 12; 13; 14; 15; 16; 17; 18; 19; 20; 21;
+ 22; 23; 24; 25; 26; 27; 28; 29; 30; 31; 32; 33; 34; 35; 36; 37; 38; 39; 40;
+ 41; 42]
+```
+
+The `Seq` module also has a function `Seq.filter`:
+```ocaml
+# Seq.filter;;
+- : ('a -> bool) -> 'a Seq.t -> 'a Seq.t = <fun>
+```
+It builds a sequence of elements satisfying a condition.
+
+Using `Seq.filter`, it is possible to make a straightforward implementation of the
+[Sieve of Eratosthenes](https://en.wikipedia.org/wiki/Sieve_of_Eratosthenes).
+Here it is:
+```ocaml
+let rec sieve seq () = match seq () with
+  | Seq.Cons (m, seq) -> Seq.Cons (m, sieve (Seq.filter (fun n -> n mod m > 0) seq))
+  | seq -> seq
+let facts = ints_from 2 |> sieve;;
+```
+
+This code can be used to generate lists of prime numbers. For instance, here is
+the list of 100 first prime numbers:
+```ocaml
+# facts |> take 100 |> List.of_seq;;
+- : int list =
+[2; 3; 5; 7; 11; 13; 17; 19; 23; 29; 31; 37; 41; 43; 47; 53; 59; 61; 67; 71;
+ 73; 79; 83; 89; 97; 101; 103; 107; 109; 113; 127; 131; 137; 139; 149; 151;
+ 157; 163; 167; 173; 179; 181; 191; 193; 197; 199; 211; 223; 227; 229; 233;
+ 239; 241; 251; 257; 263; 269; 271; 277; 281; 283; 293; 307; 311; 313; 317;
+ 331; 337; 347; 349; 353; 359; 367; 373; 379; 383; 389; 397; 401; 409; 419;
+ 421; 431; 433; 439; 443; 449; 457; 461; 463; 467; 479; 487; 491; 499; 503;
+ 509; 521; 523]
+```
+
+The function `sieve` is recursive, in OCaml and common senses: it is defined
+using the `rec` keyword and calls itself. However, some call that kind of
+function [_corecursive_](https://en.wikipedia.org/wiki/Corecursion). This word
+is used to emphasize that, by design, it does not terminate. Strictly speaking,
+the sieve of Eratosthenes is not an algorithm either since it does not
+terminate. This implementation behaves the same.
+
+## Unfolding Sequences
+
+Standard higher-order iteration functions are available on sequences. For
+instance:
+* `Seq.iter`
+* `Seq.map`
+* `Seq.fold_left`
+
+All those are also available for `Array`, `List` and `Set` and behave
+essentially the same. Observe that there is no `fold_right` function. Since
+OCaml 4.11 there is something which isn't (yet) available on other types:
+`unfold`. Here is how it is implemented:
+```ocaml
+let rec unfold f seq () = match f seq with
+  | None -> Nil
+  | Some (x, seq) -> Cons (x, unfold f seq)
+```
+And here is its type:
+```ocaml
+val unfold : ('a -> ('b * 'a) option) -> 'a -> 'b Seq.t = <fun>
+```
+Unlike previously mentioned iterators `Seq.unfold` does not have a sequence
+parameter, but a sequence result. `unfold` provides a general means to build
+sequences. For instance, `Seq.ints` can be implemented using `Seq.unfold` in a
+fairly compact way:
+```ocaml
+let ints = Seq.unfold (fun n -> Some (n, n + 1));;
+```
+
+As a fun fact, one should observe `map` over sequences can be implemented using
```

Comment:
```suggestion
As a fun fact, one should observe `map` over sequences, as it can be implemented using
```
Currently, the sentence loses some clarity in the middle. I took a stab at what I think you meant, but if it's incorrect, clarify it for me and I'll fix it.